test/test_handlers.py

```diff
@@ -235,6 +258,102 @@ def test_plate(model):
+            assert subsample_data.shape == (subsample_size,)
+
+
+def test_subsample_param():
+    data = jnp.arange(100.)
+    subsample_size = 7
+    with handlers.seed(rng_seed=0):
+        with numpyro.plate("a", len(data), subsample_size=subsample_size):
+            p0 = numpyro.param("p0", 0., event_dim=0)
+            assert jnp.shape(p0) == ()
+            p = numpyro.param("p", 0.5 * jnp.ones(len(data)), event_dim=0)
+            assert len(p) == subsample_size
+
+
+def test_subsample_substitute():
+    data = jnp.arange(100.)
+    subsample_size = 7
+    subsample = jnp.array([13, 3, 30, 4, 1, 68, 5])
+    with handlers.trace() as tr, handlers.seed(rng_seed=0), handlers.substitute(data={"a": subsample}):
+        with numpyro.plate("a", len(data), subsample_size=subsample_size) as idx:
+            assert data[idx].shape == (subsample_size,)
+            assert_allclose(idx, subsample)
+    assert tr["a"]["kwargs"]["rng_key"] is None
+
+
+def test_subsample_replay():
+    data = jnp.arange(100.)
+    subsample_size = 7
+
+    with handlers.trace() as guide_trace, handlers.seed(rng_seed=0):
+        with numpyro.plate("a", len(data), subsample_size=subsample_size) as idx:
+            pass
+
+    with handlers.seed(rng_seed=1), handlers.replay(guide_trace=guide_trace):
+        with numpyro.plate("a", len(data)) as idx:
+            assert data[idx].shape == (subsample_size,)
+
+
+@pytest.mark.parametrize("scale", [1., 2.], ids=["unscaled", "scaled"])
+@pytest.mark.parametrize("subsample", [False, True], ids=["full", "subsample"])
+def test_subsample_gradient(scale, subsample):
+    data = jnp.array([-0.5, 2.0])
+    subsample_size = 1 if subsample else len(data)
+    precision = 0.06 * scale
+
+    def model(subsample):
+        with handlers.substitute(data={"data": subsample}):
+            with numpyro.plate("data", len(data), subsample_size) as ind:
+                x = data[ind]
+                z = numpyro.sample("z", dist.Normal(0, 1))
+                numpyro.sample("x", dist.Normal(z, 1), obs=x)
+
+    def guide(subsample):
+        scale = numpyro.param("scale", 1.)
+        with handlers.substitute(data={"data": subsample}):
+            with numpyro.plate("data", len(data), subsample_size):
+                loc = numpyro.param("loc", jnp.zeros(len(data)), event_dim=0)
+                numpyro.sample("z", dist.Normal(loc, scale))
+
+    if scale != 1.:
+        model = handlers.scale(model, scale=scale)
+        guide = handlers.scale(guide, scale=scale)
+
+    num_particles = 50000
+    optimizer = optim.Adam(0.1)
+    elbo = ELBO(num_particles=num_particles)
+    svi = SVI(model, guide, optimizer, loss=elbo)
+    svi_state = svi.init(random.PRNGKey(0), None)
+    params = svi.optim.get_params(svi_state.optim_state)
+    normalizer = 2 if subsample else 1
+    if subsample_size == 1:
+        subsample = jnp.array([0])
+        _, grads1 = value_and_grad(lambda x: svi.loss.loss(
```

Comment:
could we also check that the values agree, since it's not much extra work?

Reply:
Hi @fritzo, I just added a test for value agreement. But I am not sure if that is what you mentioned.